File: lifo_ai_core/database/operations.ts
import { Database } from '@/types/supabase'
import { SupabaseClient } from '@supabase/supabase-js'

type Store = Database['business']['Tables']['stores']['Row']
type Batch = Database['inventory']['Tables']['batches']['Row']
type Product = Database['inventory']['Tables']['products']['Row']
// TODO: Re-enable when global schema is ready
// type GlobalProduct = Database['global']['Tables']['products']['Row']
type GlobalProduct = any // Temporary stub
// TODO: Re-enable when store_product table is ready
// type StoreProduct = Database['business']['Tables']['store_product']['Row']
type StoreProduct = any // Temporary stub

export class InventoryOperations {
  private supabase: SupabaseClient<Database>

  constructor(supabaseClient: SupabaseClient<Database>) {
    this.supabase = supabaseClient
  }

  async validateStoreAccess(
    storeId: string,
    userId: string,
    requiredRole: string = 'staff',
  ): Promise<boolean> {
    const { data, error } = await this.supabase.rpc('user_has_store_access', {
      target_store_id: storeId,
      required_role: requiredRole,
    })

    if (error) {
      console.error('Error validating store access:', error)
      return false
    }

    return data || false
  }

  async getUserStores(userId: string): Promise<Store[]> {
    try {
      // Temporary workaround: Get stores by owner_id to avoid RLS policy recursion
      // This will be replaced once RLS policies are properly configured
      const { data: stores, error: storesError } = await this.supabase
        .schema('business')
        .from('stores')
        .select('*')
        .eq('owner_id', userId)
        .eq('is_active', true)

      if (storesError) {
        console.error('Error fetching stores:', storesError)
        // Return empty array to allow creating new store
        return []
      }

      return stores || []
    } catch (error) {
      console.error('Error in getUserStores:', error)
      return []
    }
  }

  async createStore(storeData: Partial<Store>, ownerId: string): Promise<Store> {
    try {
      const { data, error } = await this.supabase
        .schema('business')
        .from('stores')
        .insert({
          ...storeData,
          owner_id: ownerId,
          is_active: true,
          store_code: storeData.store_code ?? 'DEFAULT_CODE',
          store_name: storeData.store_name ?? 'Untitled Store',
        })
        .select()
        .single()

      if (error) {
        console.error('Error creating store:', error)
        throw error
      }

      // Add owner to store_users table
      const { error: storeUserError } = await this.supabase
        .schema('business')
        .from('store_users')
        .insert({
          store_id: data.store_id,
          user_id: ownerId,
          role_in_store: 'owner',
          permissions: {
            can_upload_inventory: true,
            can_apply_discounts: true,
            can_view_analytics: true,
          },
          assigned_by: ownerId,
        })

      if (storeUserError) {
        console.error('Error adding owner to store_users:', storeUserError)
        // Continue despite this error - owner_id field in stores table provides fallback
      }

      console.log('Store created successfully:', data.store_id)

      return data
    } catch (error) {
      console.error('Error in createStore:', error)
      throw error
    }
  }

  // =============================================
  // GLOBAL PRODUCTS OPERATIONS - TEMPORARILY DISABLED
  // =============================================

  async findGlobalProductByBarcode(barcode: string): Promise<GlobalProduct | null> {
    // TODO: Re-enable when global schema is ready
    console.warn('Global products functionality temporarily disabled')
    return null
  }

  async searchGlobalProducts(
    searchTerm: string,
    storeId?: string,
    limit: number = 20,
  ): Promise<GlobalProduct[]> {
    // TODO: Re-enable when global schema is ready
    console.warn('Global products functionality temporarily disabled')
    return []
  }

  async createGlobalProduct(productData: {
    name: string
    brand?: string
    barcode?: string
    primary_category: string
    typical_shelf_life_days?: number
    unit_type?: string
    created_by: string
  }): Promise<GlobalProduct> {
    // TODO: Re-enable when global schema is ready
    console.warn('Global products functionality temporarily disabled')
    throw new Error('Global products functionality temporarily disabled')
  }

  async addProductToStore(
    storeId: string,
    productId: string,
    pricing: {
      default_cost_price: number
      default_selling_price: number
      store_specific_sku?: string
      supplier_code?: string
    },
    userId: string,
  ): Promise<StoreProduct> {
    // TODO: Re-enable when global schema is ready
    console.warn('Global products functionality temporarily disabled')
    throw new Error('Global products functionality temporarily disabled')
  }

  async getStoreProducts(
    storeId: string,
    options: {
      page?: number
      limit?: number
      category?: string
      active_only?: boolean
    } = {},
  ): Promise<{ data: any[]; count: number }> {
    // TODO: Re-enable when global schema is ready
    console.warn('Global products functionality temporarily disabled')
    return { data: [], count: 0 }
  }

  async createBatchWithGlobalProduct(batchData: {
    global_product_id: string
    store_id: string
    batch_number: string
    expiry_date: string
    manufacture_date?: string
    initial_quantity: number
    current_quantity: number
    cost_price?: number
    selling_price?: number
    location_code?: string
    batch_source?: string
    barcode_scanned?: string
    created_by: string
  }): Promise<Batch> {
    // TODO: Re-enable when global schema is ready
    console.warn('Global products functionality temporarily disabled')
    throw new Error('Global products functionality temporarily disabled')
  }

  async processCsvBatch(
    csvData: any[],
    storeId: string,
    userId: string,
  ): Promise<{ processed: number; errors: string[] }> {
    // TODO: Re-enable when global schema is ready
    console.warn('CSV processing functionality temporarily disabled')
    return { processed: 0, errors: ['CSV processing functionality temporarily disabled'] }
  }

  async getStoreInventoryAlerts(storeId: string, threshold: number = 0.6): Promise<any[]> {
    // TODO: Re-enable when scoring system is ready
    console.warn('Inventory alerts functionality temporarily disabled')
    return []
  }

  async getStoreInventory(
    storeId: string,
    options: {
      page?: number
      limit?: number
      category?: string
      status?: string
    } = {},
  ): Promise<{ data: any[]; count: number }> {
    // TODO: Re-enable when inventory system is ready
    console.warn('Inventory functionality temporarily disabled')
    return { data: [], count: 0 }
  }

  async updateBatchQuantity(batchId: string, newQuantity: number, userId: string): Promise<void> {
    // TODO: Re-enable when inventory system is ready
    console.warn('Inventory functionality temporarily disabled')
    throw new Error('Inventory functionality temporarily disabled')
  }

  async applyDiscount(batchId: string, discountPercent: number, userId: string): Promise<void> {
    // TODO: Re-enable when inventory system is ready
    console.warn('Inventory functionality temporarily disabled')
    throw new Error('Inventory functionality temporarily disabled')
  }

  private calculateShelfLife(category: string): number {
    const shelfLifeMap: Record<string, number> = {
      fresh_produce: 3,
      fresh_meat_fish: 2,
      bakery_fresh: 2,
      dairy: 7,
      deli_prepared: 3,
      frozen: 90,
      chilled_packaged: 14,
      pantry_staples: 365,
      canned_jarred: 730,
      dry_goods: 180,
      beverages: 365,
      spices_condiments: 730,
    }

    return shelfLifeMap[category] || 30
  }

  async getStoreStats(storeId: string): Promise<{
    totalProducts: number
    totalBatches: number
    activeAlerts: number
    totalValue: number
    expiringItems: number
  }> {
    // TODO: Re-enable when all systems are ready
    console.warn('Store stats functionality temporarily disabled')
    return {
      totalProducts: 0,
      totalBatches: 0,
      activeAlerts: 0,
      totalValue: 0,
      expiringItems: 0,
    }
  }
}

File: lifo_ai_core/database/operations.py
"""
Database operations for LIFO AI Core
Python implementation of database operations for inventory management
"""

import logging
from typing import Any, Dict, List, Optional
from datetime import datetime

# Use standard logging for database operations
logger = logging.getLogger(__name__)


class DatabaseOperationError(Exception):
    """Raised when database operations fail"""
    pass


class InventoryOperations:
    """
    Python implementation of inventory database operations
    This replaces the TypeScript version for proper Python integration
    """
    
    def __init__(self, database_connection: Optional[Any] = None):
        """
        Initialize with optional database connection
        In production, this would receive a proper database connection
        """
        self.db = database_connection
        self.logger = logging.getLogger(f"{__name__}.inventory_operations")
        
    async def validate_store_access(
        self, 
        store_id: str, 
        user_id: str, 
        required_role: str = "staff"
    ) -> bool:
        """
        Validate if user has access to store with required role
        
        Args:
            store_id: Store identifier
            user_id: User identifier  
            required_role: Minimum required role
            
        Returns:
            bool: True if user has access
        """
        if not self.db:
            self.logger.warning("Database connection not available for store access validation")
            return False
            
        try:
            # Implementation would check user permissions in database
            # For now, return True for development
            self.logger.info(
                f"Store access validation - store_id={store_id}, user_id={user_id}, role={required_role}"
            )
            return True
            
        except Exception as e:
            self.logger.error(f"Store access validation failed: {e}")
            return False
    
    async def get_user_stores(self, user_id: str) -> List[Dict[str, Any]]:
        """
        Get all stores accessible by user
        
        Args:
            user_id: User identifier
            
        Returns:
            List of store dictionaries
        """
        if not self.db:
            self.logger.warning("Database connection not available for store lookup")
            return []
            
        try:
            # Implementation would query database for user stores
            # For now, return empty list
            self.logger.info(f"Getting user stores for user_id={user_id}")
            return []
            
        except Exception as e:
            self.logger.error(f"Failed to get user stores: {e}")
            return []
    
    async def create_store(
        self, 
        store_data: Dict[str, Any], 
        owner_id: str
    ) -> Dict[str, Any]:
        """
        Create a new store
        
        Args:
            store_data: Store information
            owner_id: Store owner user ID
            
        Returns:
            Created store data
            
        Raises:
            DatabaseOperationError: If creation fails
        """
        if not self.db:
            raise DatabaseOperationError("Database connection not available")
            
        try:
            # Implementation would create store in database
            store_id = f"store_{datetime.now().strftime('%Y%m%d_%H%M%S')}"
            
            created_store = {
                "store_id": store_id,
                "owner_id": owner_id,
                "store_name": store_data.get("store_name", "New Store"),
                "store_code": store_data.get("store_code", "DEFAULT"),
                "is_active": True,
                "created_at": datetime.now().isoformat()
            }
            
            self.logger.info(f"Store created - store_id={store_id}, owner_id={owner_id}")
            return created_store
            
        except Exception as e:
            self.logger.error(f"Store creation failed: {e}")
            raise DatabaseOperationError(f"Failed to create store: {e}")
    
    async def find_global_product_by_barcode(self, barcode: str) -> Optional[Dict[str, Any]]:
        """
        Find global product by barcode
        
        Args:
            barcode: Product barcode
            
        Returns:
            Product data if found, None otherwise
        """
        if not self.db:
            self.logger.warning("Database connection not available for product lookup")
            return None
            
        try:
            # Implementation would query global products table
            self.logger.info(f"Searching for product by barcode={barcode}")
            return None
            
        except Exception as e:
            self.logger.error(f"Product search failed: {e}")
            return None
    
    async def search_global_products(
        self, 
        search_term: str, 
        store_id: Optional[str] = None, 
        limit: int = 20
    ) -> List[Dict[str, Any]]:
        """
        Search global products by name, brand, or SKU
        
        Args:
            search_term: Search query
            store_id: Optional store filter
            limit: Maximum results
            
        Returns:
            List of matching products
        """
        if not self.db:
            self.logger.warning("Database connection not available for product search")
            return []
            
        try:
            # Implementation would search products table
            self.logger.info(
                f"Searching global products - query={search_term}, store_id={store_id}, limit={limit}"
            )
            return []
            
        except Exception as e:
            self.logger.error(f"Product search failed: {e}")
            return []
    
    async def create_global_product(self, product_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Create a new global product
        
        Args:
            product_data: Product information
            
        Returns:
            Created product data
            
        Raises:
            DatabaseOperationError: If creation fails
        """
        if not self.db:
            raise DatabaseOperationError("Database connection not available")
            
        try:
            # Implementation would create product in database
            product_id = f"prod_{datetime.now().strftime('%Y%m%d_%H%M%S')}"
            
            created_product = {
                "product_id": product_id,
                "name": product_data["name"],
                "brand": product_data.get("brand"),
                "barcode": product_data.get("barcode"),
                "category": product_data["primary_category"],
                "typical_shelf_life_days": product_data.get("typical_shelf_life_days", 30),
                "unit_type": product_data.get("unit_type", "pcs"),
                "created_by": product_data["created_by"],
                "sku": f"SKU-{datetime.now().strftime('%Y%m%d%H%M%S')}",
                "created_at": datetime.now().isoformat()
            }
            
            self.logger.info(f"Global product created - product_id={product_id}")
            return created_product
            
        except Exception as e:
            self.logger.error(f"Global product creation failed: {e}")
            raise DatabaseOperationError(f"Failed to create global product: {e}")
    
    def calculate_shelf_life(self, category: str) -> int:
        """
        Calculate typical shelf life for a category in days
        
        Args:
            category: Product category
            
        Returns:
            Shelf life in days
        """
        shelf_life_map = {
            "fresh_produce": 3,
            "fresh_meat_fish": 2,
            "bakery_fresh": 2,
            "dairy": 7,
            "deli_prepared": 3,
            "frozen": 90,
            "chilled_packaged": 14,
            "pantry_staples": 365,
            "canned_jarred": 730,
            "dry_goods": 180,
            "beverages": 365,
            "spices_condiments": 730,
        }
        
        return shelf_life_map.get(category.lower(), 30)
    
    async def get_store_stats(self, store_id: str) -> Dict[str, Any]:
        """
        Get comprehensive store statistics
        
        Args:
            store_id: Store identifier
            
        Returns:
            Dictionary with store statistics
        """
        if not self.db:
            self.logger.warning("Database connection not available for store stats")
            return {
                "total_products": 0,
                "total_batches": 0,
                "active_alerts": 0,
                "total_value": 0.0,
                "expiring_items": 0,
            }
            
        try:
            # Implementation would calculate real statistics from database
            self.logger.info(f"Getting store statistics for store_id={store_id}")
            
            return {
                "total_products": 0,
                "total_batches": 0,
                "active_alerts": 0,
                "total_value": 0.0,
                "expiring_items": 0,
            }
            
        except Exception as e:
            self.logger.error(f"Failed to get store stats: {e}")
            return {
                "total_products": 0,
                "total_batches": 0,
                "active_alerts": 0,
                "total_value": 0.0,
                "expiring_items": 0,
            }


# Factory function for creating operations instance
def create_inventory_operations(database_connection: Optional[Any] = None) -> InventoryOperations:
    """
    Factory function to create InventoryOperations instance
    
    Args:
        database_connection: Optional database connection
        
    Returns:
        InventoryOperations instance
    """
    return InventoryOperations(database_connection)